### DISCUSSION TITLE
Question on Laptimes and total racetime

### DISCUSSION THREAD

**Discussion by JochemZ** [theOehrly/Fast-F1#86] [Q&A] [answered]
Hi,

thanks for this great development! Is it possible to calculate to total racetime for a driver. I did it but i cannot get the same results as the real results and the results from ERGAST:

<img width="553" alt="image" src="https://user-images.githubusercontent.com/6162547/145676854-6070f938-136e-43e6-ac05-8b45f2313f2e.png">

I have a total time of 2:06:15.291
these are my results:
<img width="449" alt="image" src="https://user-images.githubusercontent.com/6162547/145676766-69f0538a-d7b3-42db-b094-2faceabe053b.png">

and these are from ERGAST
<img width="542" alt="image" src="https://user-images.githubusercontent.com/6162547/145676966-f3bdc271-5c9f-4433-835e-256b8be04220.png">

So the problem is in the first lap.

i did the same for Max Verstappen and his total time is: 02:06:22.1300000
so the Delta between VER and HAM is not the wanted 11.825 but 7.......... ( so there is a big gap)

i must doing something wrong, please help.....

<img width="510" alt="image" src="https://user-images.githubusercontent.com/6162547/145676926-ce8cb455-4b4c-4c53-974d-b671ab1edc88.png">

**Comment by theOehrly** [comment, marked as answer]
I just took a look at this. There's two things here. You say, that the first lap is the problem. Well, the first lap is not entirely accurate but it's not 5 seconds off. The first lap inaccuracies that you observe are due to the fact, that the api does not provide lap start timestamps. Therefore we don't know when the first lap started. For races, I'm using the start time of the session as the start time of the first lap. This is not entirely accurate but pretty good. The intention behind it only was to allow getting telemetry for the first lap. Because for that we need a start and end time.

Now, about the total race time. Here is how you can calculate it fairly accurately. Note that Verstappen did receive a 5 second penalty at the 2021 Saudi Arabian GP which was added to his total race time!

```
ver = laps.pick_driver('VER')
ham = laps.pick_driver('HAM')

ver_total = ver['Time'].iloc[-1] - ver['LapStartTime'].iloc[0]
# Timedelta('0 days 02:06:22.130000')
ham_total = ham['Time'].iloc[-1] - ham['LapStartTime'].iloc[0]
# Timedelta('0 days 02:06:15.291000')

ver_total = ver_total + pandas.Timedelta(seconds=5)

print(ver_total - ham_total)
# 0 days 00:00:11.839000
```
This is pretty close to the actual value. I don't know when the clock for total race time is actually started. Is it when the lights go out or when each driver crosses the start/finish line? Maybe there's some difference here.

A quicker way to get just the delta at the end, withouth calculating the total race time is 
```
ver['Time'].iloc[-1] - ham['Time'].iloc[-1] + pandas.Timedelta(seconds=5)
# Timedelta('0 days 00:00:11.839000')
```

**Comment by JochemZ** [comment]
Many, thnx! i'm not a python/pandas expert so did not know about the iloc option to select a certain row. think i will use the times from ERGAST for reporting, and your times pure for telemetry/graphs etc. Thnx Again!

**Comment by juniormoose79** [comment]
The total racetime for each driver start at the same time, which is when the lights go out.

A first lap time is not officially published for competitors as the first lap distance completed is firstly different for every driver (due to the staggered nature of the starting grid) and secondly not necessarily a full lap distance dependent on the location of the official timing line to the starting grid.  All subsequent lap times (lap 2 onwards) are therefore suitable comparative due to being "complete" laps over nominally the same distance (ignoring the driving lines/characteristics of the individual drivers)

**Comment by theOehrly** [comment]
@juniormoose79 Thank you for that information.

This just gave me an idea for how the difference between the official and our calculated total race time could potentially be used to improve data synchronization.
